### PR TITLE
Get StreamID while still in the MsQuic callback

### LINKED ISSUE #69070: [QUIC] Don't use GetParam to get stream ID
The operation is expensive on both client (5.0% incl time in HttpClient benchmark) and server (3.7% incl time) -- it seems msquic is waiting on the lock inside

- for client, it should be possible to get stream ID from stream start event
- for server, I believe the data is not available ATM, but I think it might be possible to extend a stream accept event (PEER_STREAM_STARTED.StreamId = PEER_STREAM_STARTED.Stream->Id) -- thoughts @nibanks?

### PR DESCRIPTION
**Pull request by rzikm**
Incoming connections: Call `GetParam` from `MsQuicStream` ctor (which is called from `MsQuicConnection.NativeCallback`)
Outgoing connections: Get id from `STREAM_START_COMPLETE` event

Fixes #69070.

### CODE REVIEW

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 1]
```diff
@@ -257,12 +258,7 @@ internal override long StreamId
             get
             {
                 ThrowIfDisposed();
-
-                if (_streamId == -1)
-                {
-                    _streamId = GetStreamId();
-                }
-
+                Debug.Assert(_streamId != -1);
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:261]: I assume that in both directions (incoming/outgoing), the user cannot get a QuicStream instance before `_streamId` is set, correct?

Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:261]: Correct.

**Review by ManickaP — APPROVED** [review]
LGTM, thanks.

**Comment by rzikm** [comment]
Test failure is unrelated:  System.Net.Security.Tests.WorkItemExecution on Mono

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 2]
```diff
@@ -1720,6 +1723,7 @@ internal async ValueTask StartAsync(CancellationToken cancellationToken)
             }
 
             await _state.StartCompletionSource.Task.ConfigureAwait(false);
+            _state.Stream = null;
```
Comment (CarnaViire) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:1726]: if StartCompletionSource.Task throws, we would not null out the stream; we need to check whether that's OK @rzikm 
Could you please check how it behaves in other places (set-unset links to stream or to connection in the state objects); maybe it's better to null it out in places where we complete the TCS?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 3]
```diff
@@ -112,8 +112,9 @@ internal unsafe MsQuicStream(MsQuicConnection.State connectionState, SafeMsQuicS
 
             // Inbound streams are already started
             _state.StartCompletionSource.SetResult();
-
             _state.Handle = streamHandle;
+            _streamId = GetStreamId(); // needs _state.Handle to be set
```
Comment (wfurt) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:116]: We could make it static and pass the handle to GetStreamId(). But this is fine for now.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 4]
```diff
@@ -1707,6 +1709,7 @@ internal async ValueTask StartAsync(CancellationToken cancellationToken)
             int status;
             unsafe
             {
+                _state.Stream = this;
```
[the remark this answers is not in the document]
Comment (wfurt) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:1712]: One more note that the `GetTraceId()` is trying to mimic msquic so it is easier to correlate logs. If StreamId is available, we can make it part of TraceId in addition to what we have. That would make it easier IMHO to correlate with packet captures and Kestrel.

**Pull request merged**